### DISCUSSION TITLE
Finalizer example possible bug

### DISCUSSION THREAD

**Discussion by lorenzofelletti** [kubernetes-sigs/kubebuilder#4290] [Support] [answered]
The finalizer example in the book performs a `r.Update(ctx, cronJob)` [here](https://github.com/kubernetes-sigs/kubebuilder/blob/v4.3.0/docs/book/src/cronjob-tutorial/testdata/finalizer_example.go#L71), but does not requeue in case the update is successful.

Since the code roughly does
```go
func (r *CronJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	//...

	if cronJob.ObjectMeta.DeletionTimestamp.IsZero() {
		if !controllerutil.ContainsFinalizer(cronJob, myFinalizerName) {
			controllerutil.AddFinalizer(cronJob, myFinalizerName)
			if err := r.Update(ctx, cronJob); err != nil {
				return ctrl.Result{}, err
			}
		}
	}

	// reconciliation logic
}
```

Wouldn't that code as is cause an error when the reconciliation logic attempt to update the status (e.g. [here](https://github.com/kubernetes-sigs/kubebuilder/blob/v4.3.0/docs/book/src/cronjob-tutorial/testdata/project/internal/controller/cronjob_controller.go#L258))?
Shouldn't it be better to return `ctrl.Result{Requeue: true}, nil` on success, so that the finalizer is added and the reconciler immediately retriggered (thus enabling us to fetch the latest version of the resource)?

**Comment by camilamacedo86** [comment, marked as answer]
Hi @lorenzofelletti,

The best approach would be to refactor the code of the CronJob and MultiVersion samples to follow the structure generated by the [DeployImage plugin](https://book.kubebuilder.io/plugins/available/deploy-image-plugin-v1-alpha) (which scaffolds the code to manage an image provided via the command line).

### Example:
- **Controller Implementation:** [busybox_controller.go](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-plugins/internal/controller/busybox_controller.go)
- **Controller Test:** [busybox_controller_test.go](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-plugins/internal/controller/busybox_controller_test.go#L139-L146)

**The best practice is to fetch the resource before updating it.** This avoids scenarios where an update fails due to an intermediate change, which could cause the reconciliation process to restart. Refer to the example below for guidance:

https://github.com/kubernetes-sigs/kubebuilder/blob/5f8342e45517e84a8cc8dd4eff825fe340477111/testdata/project-v4-with-plugins/internal/controller/busybox_controller.go#L101-L118

To align with the above examples, we have indeed some refactoring tasks:
- [kubebuilder/issues/4140](https://github.com/kubernetes-sigs/kubebuilder/issues/4140)
- [kubebuilder/issues/4019](https://github.com/kubernetes-sigs/kubebuilder/issues/4019)

If you wish to work on and help us improve the  CronJob and MultiVersion in order of them be exactly the same controller and controller test scaffold done by [DeployImage plugin](https://book.kubebuilder.io/plugins/available/deploy-image-plugin-v1-alpha) with only the required logic for this example on top that would be great. Help is more than welcome.

**Comment by camilamacedo86** [comment]
Furthermore we created the issue : https://github.com/kubernetes-sigs/kubebuilder/issues/4291
So, I am closing this one and I hope that you do not mind.

If you see that is required please feel free to re-open this one.
